src/nks/hcvault/provider.rs

```diff
@@ -0,0 +1,512 @@
+    /// specified in the `config` parameter. The key is made persistent and associated with the provided `key_id`.
+    ///
+    /// # Arguments
+    ///
+    /// * `key_id` - A string slice that uniquely identifies the key to be created.
+    /// * `config` - A Box containing a `ProviderConfig` object that specifies the settings for the key.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` that, on success, contains `Ok(())`, indicating that the key was created successfully.
+    /// On failure, it returns a `SecurityModuleError`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// let config = get_config("rsa").unwrap();
+    /// provider.create_key("test_rsa_key", Box::new(config.clone())).expect("Failed to create RSA key");
+    /// ```
+    #[instrument]
+    fn create_key(&mut self, key_id: &str, config: Box<dyn ProviderConfig>) -> Result<(), SecurityModuleError> {
+        let mut key_length: Option<KeyBits> = None;
+        if let Some(nks_config) = config.as_any().downcast_ref::<NksConfig>() {
+            let runtime = Runtime::new().unwrap();
+            let get_and_save_keypair_result = runtime.block_on(get_and_save_key_pair(
+                &*nks_config.nks_token.clone(),
+                key_id,
+                match nks_config.key_algorithm.clone() {
+                    AsymmetricEncryption::Rsa(rsa) => {
+                        key_length = Some(rsa);
+                        "rsa"
+                    },
+                    AsymmetricEncryption::Ecc(EccSchemeAlgorithm::EcDh(EccCurves::Curve25519)) => "ecdh",
+                    AsymmetricEncryption::Ecc(_) => "ecdsa",
+                },
+                key_length,
+                Url::parse(&nks_config.nks_address).unwrap()
+            ));
+            match key_length {
+                Some(kb) => println!("XXXXX Key length: {}", u32::from(kb)),
+                None => println!("XXXXX Key length: None"),
+            }
+            match get_and_save_keypair_result {
+                Ok((result_string, new_token)) => {
+                    println!("Key pair generated and saved successfully.");
+                    self.secrets_json = Some(result_string.parse().unwrap());
+                    //safe token to config
+                    let config = NksConfig::new(
+                        new_token.clone(),
+                        nks_config.nks_address.clone(),
+                        nks_config.key_algorithm.clone(),
+                        nks_config.hash.clone(),
+                        nks_config.key_usages.clone(),
+                    );
+                    self.config = Some(config);
+                    //save token in token.json for persistence
+                    let token_data = json!({
+                     "user_token": new_token.clone()
+                     });
+                    fs::write("token.json", token_data.to_string()).expect("Error writing to token.json");
```

Comment:
This is a rather sensitive file, considering that it contains a token that can be used for Vault access. Would it be possible to refactor code to make it possible not to write tokens to a json file?

Reply:
This is intended as abstract functionality because the user needs to authenticate themselves to the backend somehow. For example, after the keys are created, to retrieve them again.
The location where this file is saved would need to be changed according to the use case of the code, For instance, this file could be saved in the encrypted part of an app or something similar.